sparse/coo.py

```diff
@@ -2034,113 +2006,125 @@ def _grouped_reduce(x, groups, method, **kwargs):
     return result, inv_idx, counts
 
 
-def _elemwise_binary(func, self, other, *args, **kwargs):
-    check = kwargs.pop('check', True)
-    self_zero = _zero_of_dtype(self.dtype)
-    other_zero = _zero_of_dtype(other.dtype)
+def elemwise(func, *args, **kwargs):
+    """
+    Apply a function to any number of arguments.
 
-    func_value = func(self_zero, other_zero, *args, **kwargs)
-    func_zero = _zero_of_dtype(np.dtype(func_value))
-    if check and func(self_zero, other_zero, *args, **kwargs) != func_zero:
-        raise ValueError("Performing this operation would produce "
-                         "a dense result: %s" % str(func))
+    Parameters
+    ----------
+    func : Callable
+        The function to apply. Must support broadcasting.
+    args : tuple, optional
+        The arguments to the function. Can be :obj:`scipy.sparse.spmatrix` objects,
+        :obj:`SparseArray` objects or :obj:`numpy.ndarray` objects.
+    kwargs : dict, optional
+        Any additional arguments to pass to the function.
 
-    if not isinstance(self, COO):
-        if not check or np.array_equiv(func(self, other_zero, *args, **kwargs), func_zero):
-            return _elemwise_binary_self_dense(func, self, other, *args, **kwargs)
-        else:
-            raise ValueError("Performing this operation would produce "
-                             "a dense result: %s" % str(func))
+    Returns
+    -------
+    COO
+        The result of applying the function.
 
-    if not isinstance(other, COO):
-        if not check or np.array_equiv(func(self_zero, other, *args, **kwargs), func_zero):
-            temp_func = _reverse_self_other(func)
-            return _elemwise_binary_self_dense(temp_func, other, self, *args, **kwargs)
-        else:
+    Raises
+    ------
+    ValueError
+        If the operation would result in a dense matrix, or if the operands
+        don't have broadcastable shapes.
+
+    See Also
+    --------
+    :obj:`numpy.ufunc` : A similar Numpy construct. Note that any :code:`ufunc` can be used
+        as the :code:`func` input to this function.
+    """
+    # Because we need to mutate args.
+    args = list(args)
+    posargs = []
+    pos = []
+    for i, arg in enumerate(args):
+        if isinstance(arg, scipy.sparse.spmatrix):
+            args[i] = COO.from_scipy_sparse(arg)
+        elif isscalar(arg) or (isinstance(arg, np.ndarray)
+                               and not arg.shape):
+            # Faster and more reliable to pass ()-shaped ndarrays as scalars.
+            if isinstance(arg, np.ndarray):
+                args[i] = arg[()]
+
+            pos.append(i)
+            posargs.append(args[i])
+        elif isinstance(arg, SparseArray) and not isinstance(arg, COO):
+            args[i] = COO(arg)
```

Comment:
This line doesn't get hit by tests.  Should we add a small DOK test?

Reply:
Added!